C backend

vadz opened the pull request:
After 2016 and 2019, I'm trying to return one more/last time to this backend and finally bring it up to the state in which it will be useful enough to merge. It is still far from being there yet though...

vadz commented:
I think this could be considered for merging now. There are definitely a lot of remaining problems and the existing code could be improved, but it does work for wrapping a pretty big API using different standard containers, shared pointers etc (but not using multiple inheritance, C arrays and other unsupported, at least at C++ level, features) and so IMO would be already useful to have.

Note that this relies on/include #2095 and #2096.

vadz commented:
I know there are a lot of changes here, but most of them are in the new files created by this branch and most of the changes to the existing files are trivial. Is there any chance that this gets reviewed and merged in the observable future?

ojwb commented on .github/workflows/ci.yml:
```diff
@@ -22,9 +22,8 @@ jobs:
-    # can define the additional "desc" field with any additional information to include in the name.
-    name: ${{ matrix.SWIGLANG || 'none' }}${{ matrix.PY3 }} ${{ matrix.ENGINE}} ${{ matrix.VER }} ${{ matrix.SWIG_FEATURES }} ${{ (matrix.compiler || 'gcc') }}${{ matrix.GCC }} ${{ matrix.CPPSTD }} ${{ matrix.CSTD }} ${{ matrix.desc }} ${{ matrix.continue-on-error && '(can fail)' }}
+    # By default, the name of the build is the language used and SWIG options.
+    name: ${{ matrix.SWIGLANG || 'none' }}${{ matrix.PY3 }} ${{ matrix.ENGINE}} ${{ matrix.VER }} ${{ matrix.SWIG_FEATURES }} ${{ (matrix.compiler || 'gcc') }}${{ matrix.GCC }} ${{ matrix.CPPSTD }} ${{ matrix.CSTD }} ${{ matrix.continue-on-error && '(can fail)' }}
```
This dropping of `desc` doesn't seem related to adding a C backend.  It also seem a useful feature, though TBH I've barely looked at the new CI stuff.

vadz replied:
Thanks, this looks like a bad merge, I'll fix this.

vadz replied:
Ah, sorry, no, I remember now: I've added `desc` to use it for anything not covered by the other variables in the build descriptions but @wsfulton removed all uses of it in 04a6ad3bb582c15d7dc13c86b9657038adeb07ce and e74b9830bbb0a3f32d044e56b671b7685cd253bf, so I've decided to remove it here too. We can keep it, of course, but to me it seems confusing to have a never used variable.

ojwb commented on Examples/test-suite/arrays_dimensionless.i:
```diff
@@ -9,10 +9,14 @@
+    static CC ccs[];
 };
 int Bar::ints[] = {700, 800, 900};
+CC Bar::ccs[] = {CC(), CC()};
```
Just noting that this isn't specific to the new backend.  It looks OK to me, but highlighting for anyone else reviewing.

vadz replied:
This was added back in a2dc2756c (Several major fixes for: arrays, static members, member func.ptrs., exceptions, ... Lots of tests runs ok now., 2009-04-13) for completely unknown reasons. I didn't revert it because I didn't see any special reason to do it, but I don't have any special reason to keep it neither and could revert it if desired.

ojwb commented on Examples/test-suite/c/operator_overload_runme.c:
```diff
@@ -0,0 +1,26 @@
+
+#include "operator_overload/operator_overload_wrap.h"
+
+#define assert(x,msg) if (!x) { printf("%d: %s\n", x, msg); exit(1); }
```
This evaluates `x` twice which is not good if it has side-effects.

It also lacks parentheses so this assert doesn't fail (because the condition evaluates like `((!1) == 2)`):

```
assert(1==2, "oops");
```

It also doesn't swallow a following `;` which can give surprising control flow in some cases (and confusing compilation errors in others) if `assert` is used in an unbraced block.

Why don't we use `<assert.h>` here rather than reinventing a rather buggy alternative version?

The standard `assert()` doesn't directly support a message, but one can just put a comment before it, which should be fine here as the testcases are not huge programs.

Or there are common hacks like the one noted at https://en.cppreference.com/w/cpp/error/assert:

```
assert(("There are five lights", 2 + 2 == 5));
```

You'd probably need to cast the string literal to `void` to avoid compiler warnings, but something like this works:

```
#include <assert.h>

#define assert2(x, msg) assert(((void)msg, x))
```

E.g. with GCC that gives me a warning free compilation with `-Wall -W -O2` and runtime output of:

```
$ ./a.out 
a.out: badassert.c:9: main: Assertion `((void)"boo", 1==2)' failed.
Aborted
```

With clang it's very similar:

```
$ ./a.out 
a.out: badassert.c:9: int main(): Assertion `((void)"boo", 1==2)' failed.
Aborted
```

ojwb replied:
I see that most of the other testcases just use the standard assert, so probably just doing that here too makes most sense.

vadz replied:
I have no idea why does this test, and it only, use a custom `assert` macro, but I'll get rid of it.

ojwb commented on Examples/test-suite/c_backend_cpp_exception.i:
```diff
@@ -0,0 +1,25 @@
+  void throwSomeUnknownException(void) {
+    throw SomeUnkownException();
+  }
+%}
```
A number of these new files lack a newline on the final line, which is probably better avoided.

vadz replied:
AFAICS (or, rather, `rg -l '[^\n]\z'` can see) there are only 4 of such files added on this branch. I can fix them, of course, but there are a bunch of existing files in master with the incomplete last line and I think it might make more sense to fix all of them at once after merging.

ojwb commented on Examples/test-suite/dynamic_cast.i:
```diff
@@ -17,14 +17,14 @@ public:
 #endif
 
-#if defined(SWIGJAVA)
+#if defined(SWIGJAVA) 
```
Random trailing whitespace added.

vadz replied:
Added in the same a2dc2756c8, I'll revert it too.

ojwb commented on Lib/c/std_except.i:
```diff
@@ -0,0 +1,34 @@
+%typemap(throws) std::range_error       "SWIG_JavaThrowException(jenv, SWIG_JavaIndexOutOfBoundsException, $1.what());\n return $null;"
+%typemap(throws) std::runtime_error     "SWIG_JavaThrowException(jenv, SWIG_JavaRuntimeException, $1.what());\n return $null;"
+%typemap(throws) std::underflow_error   "SWIG_JavaThrowException(jenv, SWIG_JavaArithmeticException, $1.what());\n return $null;"
+*/
```
This looks like a big commented out cut and paste from java library files.

vadz replied:
Removed too.

ojwb commented on Source/Include/swigwarn.h:
```diff
@@ -255,6 +255,11 @@
 /* please leave 740-759 free for Python */
 
+#define WARN_C_TYPEMAP_CTYPE_UNDEF            760
+#define WARN_C_UNSUPPORTTED                   779
```
Normally we allocate from the first reserved value up, rather than from both ends of the reserved range in.

vadz replied:
I think the intention was to clearly reserve the entire range, is this really a problem?

ojwb replied:
I guess the real problem here is that each such tiny little arbitrary difference between target languages adds cognitive load and in combination they make it harder to get your head round SWIG in its entirety, and we really need more people who can deal with stuff across SWIG rather than specific to one target language (we really need more of the latter too, but there's a particular shortage of active people in the first category).  Consistency is useful tool for taming complexity.

So if all the existing target languages do something one way, I think there ought to be a really good reason for a new language doing it differently, even if there's logic behind both approaches.

vadz replied:
I don't want to relaunch all the CI builds just for this, but if I make any other fixes later or if it's really the only thing which prevents this PR from being merged, I'll change the value of `WARN_C_UNSUPPORTTED` to 761.

wsfulton replied:
Use [skip ci] or variations in commit message to avoid rerunning CI, see https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs.

ojwb commented on Source/Modules/lang.cxx:
```diff
@@ -1320,16 +1320,21 @@ int Language::staticmemberfunctionHandler(Node *n) {
+	Append(cname, Getattr(defaultargs ? defaultargs : n, "sym:overname"));
+      } else {
+	Append(cname, "__SWIG");
+      }
```
If I follow, the only functional change here is to append `__SWIG` if `(!Getattr(n, "sym:overloaded") && code)` - the hunk below is only refactoring to hoist the `code` check to be around all this.

This seems to have potential to affect other target languages - does `(!Getattr(n, "sym:overloaded") && code)` never hold for existing target languages?

ojwb replied:
Oh this seems to be related to #2095

wsfulton replied:
#2095 is now merged to master.

ojwb commented:
`Examples/c/exception/example.cxx` is an empty file (can't seem to comment on it directly in the github UI) - is that intended?

ojwb commented on Examples/Makefile.in:
```diff
@@ -1350,6 +1350,72 @@ r_clean:
+c_cpp: $(SRCDIR_SRCS)
+	$(SWIG) -c++ -c $(SWIGOPT) -o $(ICXXSRCS) $(INTERFACEPATH)
+	$(CXX) -c $(CCSHARED) -I$(SRCDIR) $(CXXFLAGS) $(ICXXSRCS) $(SRCDIR_CXXSRCS) $(INCLUDES)
+	$(CXX_LDSHARED) $(CFLAGS) $(OBJS) $(IOBJS) $(LIBS) $(CPP_DLLIBS) -o $(CLIBPREFIX)$(TARGET)$(C_SO)
```
I would suggest prefixing this and the `C_LDSHARED` command above with `$(COMPILETOOL)`, which seems to be consistent with how that's used with other target languages.

Then the change in `Examples/test-suite/common.mk` to override `C_LDSHARED=true` etc wouldn't be necessary.

vadz replied:
Thanks!

ojwb commented on Examples/Makefile.in:
```diff
@@ -1627,3 +1693,4 @@ d_clean:
 	rm -f *_wrap* *~ .~* $(RUNME) $(RUNME).exe `find . -name \*.d | grep -v $(RUNME).d`
 	rm -f core @EXTRA_CLEAN@
 	rm -f *.@OBJEXT@ *@SO@
+
```
Extra blank line added at end of file.

vadz replied:
This one must be due to some merge weirdness because git-blame in the branch attributes it to e0067ea99 (Merge trunk rev 13025 (swig-2.0.6) into gsoc2008-maciekd, 2012-05-07)... Anyhow, I've removed it now, thanks.

ojwb commented on Examples/c/class/example.h:
```diff
@@ -0,0 +1,39 @@
+
+
+
+  
```
This looks like a copy of an old version of this example from another target language.  7+ years ago I cleaned these up to be consistent between languages (b115c984a97e5f20e9ad5979f1ad606fff549bc3 for example, but there are probably other commits) but here I notice at least `void` parameter lists and a superfluous `;` after `Shape`'s destructor's definition are back.

I realise this is due to how long-lived this branch has been, but it'd be good to compare the C examples with the equivalents from other target languages and avoid reintroducing these ancient inconsistencies.

vadz replied:
I've removed gratuitous differences between this file and the C# version, thanks.

ojwb replied:
This is a wider issue - e.g. in master:

```
$ git grep 'Circle::area' Examples/*/class/example.cxx
Examples/csharp/class/example.cxx:double Circle::area() {
Examples/d/class/example.cxx:double Circle::area() {
Examples/guile/class/example.cxx:double Circle::area() {
Examples/java/class/example.cxx:double Circle::area() {
Examples/javascript/class/example.cxx:double Circle::area() {
Examples/lua/class/example.cxx:double Circle::area() {
Examples/mzscheme/class/example.cxx:double Circle::area() {
Examples/octave/class/example.cxx:double Circle::area() {
Examples/perl5/class/example.cxx:double Circle::area() {
Examples/php/class/example.cxx:double Circle::area() {
Examples/python/class/example.cxx:double Circle::area() {
Examples/r/class/example.cxx:double Circle::area() {
Examples/ruby/class/example.cxx:double Circle::area() {
Examples/scilab/class/example.cxx:double Circle::area() {
Examples/tcl/class/example.cxx:double Circle::area() {
```

But the version added for C here has `void` here (and there are also other long-fixed niggles reintroduced, such as the comment at the top saying the file is `example.c` when it's actually `example.cxx`).

I was suggesting comparing **all** the C examples with those from other target languages (you can probably just pick one that's in a reasonable state) e.g. something like: `diff -ru --exclude=Makefile --exclude=check.list Examples/csharp Example/c` to eliminate these sort of inconsistencies.

vadz replied:
I could have missed something because it wasn't simple to see the relevant differences among all the other ones, but I think I've got rid of all such problems now (see 4748e4f3798188c07bcd7b3a0a9ab159390d5169).

vadz commented:
> `Examples/c/exception/example.cxx` is an empty file (can't seem to comment on it directly in the github UI) - is that intended?

Yes, it needs to be there for the recipes in the makefile to work. Makefiles could be made more flexible, of course, but it doesn't cost anything to keep this empty file, so I'd rather not touch them.

Thanks for the review, I'm pushing a few commits addressing the issues above.

wsfulton commented:
I'm going to read C.html first. Is this completely up to date and accurate wrt to the implementation?

vadz commented:
> I'm going to read C.html first. Is this completely up to date and accurate wrt to the implementation?

I've tried to keep it up to date by documenting everything I added there, so it should be, yes. I'm not too happy with its existing organization, but I didn't want to change it and, in any case, I'd strongly prefer to wait until #1844 is merged before working more on the docs. Writing HTML manually is just not fun at all.

ojwb commented:
I think it would make sense at this point to just get this merged to master, and handle cleaning up the docs and other minor issues as follow-up PRs.  Maintaining branches for large changes for a long time adds tedious work dealing with merge conflicts and risks regressing global clean up changes made to master because instances can remain on the branch.

It's also morale-sapping to have to wonder if a branch will ever actually get merged.  Getting it merged probably also makes it more likely we'll get contributions to it from others.  If it's in 4.1.0 and marked as "experimental" (or whatever that status is called) then the bar to trying it out is reduced, and if it's in a release people are more likely to invest time in trying it compared to a branch that's been going since 2013 (I think that's right, it's before 2016 anyway).

@vadz isn't a drive-by contributor - we don't need to worry about him vanishing once this is merged.

wsfulton commented:
I have actually been quietly combing through it since my last response and unfortunately have come down with Covid. I'm getting a whole view before responding, but my impression to date is quite good, but may have some questions if I can't answer them myself once I've digested it all. The lack of up to date docs is slowing me down. I'll let you know when I've finished looking. The only thing pre-merge that I've spotted to date is to fix the 'in-source' examples don't work. Also I spotted that C target language support is missing from testflags.py and c8645aab9ec227f5436bf8a061cc7a20f5a47717 will highlight that.

vadz commented:
> I have actually been quietly combing through it since my last response and unfortunately have come down with Covid.

Sorry to hear this, hopefully you get better soon.

> I'm getting a whole view before responding, but my impression to date is quite good, but may have some questions if I can't answer them myself once I've digested it all. The lack of up to date docs is slowing me down.

There are some hopefully useful comments in the code and, of course, you could always just ask if you have any questions and I'd do my best to answer them (but it would be better to do it soon because I will have forgotten everything about this code myself in a couple more months).

Concerning the docs, I don't think C backend docs are significantly out of line with many other similar backends (e.g. D). It doesn't mean that they couldn't be improved, of course, but it would be really helpful if we could migrate the docs to ReST (or Markdown or AsciiDoctor or whatever that isn't raw HTML -- this is just too painful to write) before doing it.

> I'll let you know when I've finished looking. The only thing pre-merge that I've spotted to date for fixing is the 'in-source' examples don't work. Also I spotted that C target language support is missing from testflags.py and [c8645aa](https://github.com/swig/swig/commit/c8645aab9ec227f5436bf8a061cc7a20f5a47717) will highlight that.

Adding "c" to `testflags.py` is trivial, of course, and I can do that or this could be also done when merging. And I can look at fixing in-tree examples build, but I admit that it's hard to remain motivated to fix such minor problems. In addition to what Olly wrote, I currently also have JS changes and Python stable API ones, as well as a dozen of other more minor things, in my own branch and I think that by now I'm almost resigned to have to maintain it myself for the observable future, as it seems that none of these things is going to be merged into the official SWIG.

wsfulton commented:
I do wholeheartedly apologise for causing the lack of motivation. There is just a lot of complexity in SWIG and so much to do and I try only to respond to pull requests when I know I can usefully contribute and focus on them. Large pull requests are actually quite stressful and daunting, this one in particular as it is such an important language module. I'll try remedy and am focusing on finishing up my review of it now this weekend while I'm stuck indoors.

Where would you prefer to discuss post-merge topics and TODO lists for this module? It is a really important module and worthy of a thread in itself. We could use swig-devel or try the wiki or the github conversations.

The documentation is a source of immense frustration to me and is being discussed on other threads, but html is what we have. Would you be comfortable putting text inside of `<pre></pre>` elements as you get to see the text as you typed it. I'm happy to convert to proper html, I don't find that time consuming. Use markup or whatever format you prefer inside of the elements if you like. The content is really valuable compared to the formatting of it.

vadz commented:
William, I understand perfectly well the burden of being a maintainer, so I don't blame you for anything but, in particular, I also understand that nothing is ever perfect and there must be some compromise between the quality and speed of development. Where should the line be exactly is a complicated question, but to me it seems clear that in SWIG case it skews too far towards stability at the expense of new functionality. And I'm afraid that neither of us is going to have more time or suddenly become much more productive with what time we do have, so we need to do what we can with what we've got and for me this means integrating the changes sooner even if they're not perfect -- it's always much simpler to improve them later on master than keeping working on some other branch for many months, if not years.

Anyhow, these philosophical issues aside, I'd be glad to discuss this wherever is the most convenient for you. Swig-devel is perfectly fine, but just about anywhere else would be too.

Concerning the documentation, I was thinking of giving more examples of SWIG input and output for this backend, and this is the worst thing to format in raw HTML. I really hope the migration to ReST can be completed soon and, in fact, I was planning to look at finishing it myself if/when this one gets merged. But until this gets done writing any new docs just seems like a waste of time.

Again, just to be clear: I know the code in this backend is far from perfect, but unfortunately the choice is not between using the current version and making it perfect, but rather between using the current imperfect but more or less working version and not having anything at all.

wsfulton commented:
> Concerning the documentation, I was thinking of giving more examples of SWIG input and output for this backend, and this is the worst thing to format in raw HTML.

While escaping some characters is technically needed, modern browsers show them just fine within preformatted elements `<pre></pre>` and I'm happy to correct this for you while we have to work in raw html. Just add new blocks of text, examples and all within the suggested preformatted elements as if you are writing plain text documentation. Think of this as draft rendering for sorting out later. I really value documentation highly, it's as important as the code because without it, it creates a big burden on users and our small development community. What you know is invaluable and saves me time so, please, please just write down what you know! We have raw html doc replacement discussions on other threads, I'd rather focus on the documentation content for C in this thread.

Don't worry about the code not being perfect in the C backend, it's an experimental language. The merge just needs to conform to the requirements for being an experimental language and all my comments to date are around that.

vadz commented:
What exactly would you like to see documented? It would be nice to have some exact goals in mind.

wsfulton commented:
Your suggestion of giving more examples of SWIG input and output for this backend sounds great. Python and Java chapter are our best role models for converting to C, but consider that a longer term goal. I think the best thing would be to read through the current docs and fix what you know is incorrect or poor in the first instance.

wsfulton commented:
I'll save the majority of my questions for post-merge, but here is one:

Could you please summarise briefly the history since 3d6880aad18dea66f778033634381f02373f4e7b wrt to the C++ layer. If I recall correctly in the two GSoC projects, we were modelling all this on the usual SWIG approach: a flat C layer and another (C++ in this case) proxy class layer wrapper. It seems that the proxy layer is still there but not called proxy now??? Did the proxy layer make a comeback after that commit in some evolved, improved way? Are you using another name for the proxy layer?

I wonder if 'C target language' is the right name, given it outputs more C++ than C!

ojwb commented:
My memory (though I might be wrong) is these were two separate GSoC project ideas - one which was worked on in 2008 and again in 2012 to take a C++ library and produce a flat C wrapping of it (that's this one):

https://github.com/swig/swig/wiki/GSoC-2008-C-Backend
https://sourceforge.net/p/swig/news/2012/11/summer-of-code-2012/

And a separate idea to create a C++ class wrapping around a C API (for example one which had a "pass a structure as the first parameter to functions which operate on it" design).  I'm not entirely sure if the second one ever actually became a project, but I found students saying they were interested in it.

wsfulton commented:
Yes, there were two GSoC projects, the second building on top of the first. From what I can make out, at the end the second (Leif's work), the output had a proxy layer, but this layer was purely a flattened api. It was C only. The C++ proxy classes were not yet in existence. It seems now that the C proxy layer has been slimmed down and is more tightly coupled to the original C layer. It looks like this module has morphed into providing two target languages now, a C layer, plus a C++ layer. This was the original goal. Although, I don't know that we ever thought about combining the C++ layer directly into the 'C target language', it does works and I think it's okay. There is thus a fundamental question, is this now a C target language or both C and C++?!?

And that leads me to one further question that I've been attempting to figure out as it's somewhat blurred. What is the intended approach to testing this in the examples and test-suite? Given that there are two outputs, C and C++, it seems that the testing is a sort of mix of C and C++ and it's not entirely clear to me whether C is actually working for the entire test-suite as the C++ compiler is used in many instances on the output. There are both runme.c and runme.cxx files. I'll keep reverse engineering, but would be helpful to get a fundamental overview @vadz and please correct anything I've got wrong above.

vadz commented:
I don't remember everything myself neither, but I think there was some confusion about what the proxy layer is since the very beginning (i.e. since 2008). At that time there was no support for generating C++ at all, but there were already things called "proxy something" in the code and AFAICS this didn't make any sense -- there is no object layer in C, so what we had was basically a single layer more or less arbitrarily split into different parts. During the second GSoC project (2012) this structure was preserved because it didn't seem fair to ask one student to correct the mistakes of another one, but I finally (mostly) got rid of it to simplify things when I tried to resurrect it in 2016 (see e.g. 806f79de147b128240bda4b34c9f11c4d42a6fb0 commit message).

But now there is indeed another layer on top of the C one, which generates C++ proxies. I didn't call it "proxy" because the term is rather overloaded in SWIG and so using it would be confusing, IMHO. I also didn't separate it into a separate language module, but this was mostly because it would have been much more work for not much gain: I'm pretty sure that C module will mostly be used for C++ wrappers generation, as this is the use case which really makes sense (whatever the docs say, nobody is going to use this module just to apply contracts to the existing C functions).

As for the tests, they actually use both C and C++, i.e. C compiler _is_ used for all the tests. C++ compiler is not, however, see a long list of `xxx.cpptest` for which `SWIG_NOCXX` is set in `Examples/test-suite/c/Makefile.in` due to lack of support for various C++ features at C++ wrappers level (but these features are supported at C level, for those that aren't supported even there we use `FAILING_CPP_TESTS`, just as in the other backends). Using `runme.c` and `runme.cxx` was the best solution I found for testing both the generated C and C++ wrappers. Another alternative would be to have just a single file but compile it using both C and C++ compilers and use `#ifdef __cplusplus` inside it, but using 2 files seems cleaner.

wsfulton commented:
It'll be interesting to see who uses it for just C, I think it could be more popular than C++ because C is much more portable than C++ and just about every other language has an interface into C and don't support C++.

I'd also be interested in your use case for the module. Could you document this please. I have a few use cases in mind around mixing compilers, which I'll add into the docs afterwards.

Even if the C and C++ layers generated come from the one module, I think longer term we should provide a C test-suite and a C++ test-suite in different directories. That'll make it easier to understand and maintain going forwards. 

If proxy classes is not the term to be used, I hope we can find and consistently use another name to distinguish the C vs C++ generated layers and use it consistently in c.cxx too and the docs. I think the term "wrapper" is even more overloaded in SWIG than proxy, so I don't know what terminology should be! I'd also like to see the C layer free of C++ generated code (except some non-trivial stuff maybe).

Feel free to merge this branch to master whenever you like. The following are prerequisites for accepting the module in master, but you can implement them before merging or shortly after, but definitely before we release 4.1.0. These are the areas where I feel it does not meet the 'experimental language' criteria:

1. In source examples don't compile.
2. Documentation is not reasonably sufficient and not always correct, as discussed.
3. Github Actions ci.yml file needs to be marked continue-as-error like the other 2 experimental languages.

Finally, I can see you've put a lot of work into this, great work. I apologise again that I havn't been able to look at it sooner. Once you have merged it, I'll do a clean up and have some boring things to commit mainly around  inconsistencies to the rest of SWIG. I may even try and help improve the functionality if I get some time.

vadz commented:
> I'd also be interested in your use case for the module. Could you document this please.

This is already documented, actually: I want to use to avoid problems due to lack of C++ ABI stability, i.e. allow using a C++ library compiled with one (version of) C++ compiler from the main program compiled using a different (version of the) C++ compiler.

> Even if the C and C++ layers generated come from the one module, I think longer term we should provide a C test-suite and a C++ test-suite in different directories. That'll make it easier to understand and maintain going forwards.

Personally I'm only interested in C++ and as it's (or should be) a superset of C, I think  it would be enough to have C++ tests.

> If proxy classes is not the term to be used, I hope we can find and consistently use another name to distinguish the C vs C++ generated layers and use it consistently in c.cxx too and the docs. I think the term "wrapper" is even more overloaded in SWIG than proxy, so I don't know what terminology should be! I'd also like to see the C layer free of C++ generated code (except some non-trivial stuff maybe).

This is what you get with `-nocxx` option.
 
> Feel free to merge this branch to master whenever you like. The following are prerequisites for accepting the module in master, but you can implement them before merging or shortly after, but definitely before we release 4.1.0. These are the areas where I feel it does not meet the 'experimental language' criteria:
> 
>     1. In source examples don't compile.

I'll try to fix this.
 
>     2. Documentation is not reasonably sufficient and not always correct, as discussed.

I'm not aware of anything really incorrect in it, do you have something precise in mind? I'll try to improve it as much as it pains me to do it in HTML (and no, I don't think that writing it in Markdown and using your as Markdown to HTML converter is an efficient use of your time neither, although thanks for offering to do it).

>     3. Github Actions ci.yml file needs to be marked continue-as-error like the other 2 experimental languages.

I'd really like to avoid it. I am not going to notice if it breaks with this flag, but I'd very much like to know about it.

ojwb commented:
> > 3. Github Actions ci.yml file needs to be marked continue-as-error like the other 2 experimental languages.
> 
> I'd really like to **not** do this. I am not going to notice if it breaks with this flag, but I'd very much like to know about it.

We don't want experimental languages with shonky testsuites which fail depending on phase of the moon causing CI runs to fail, but the experimental status really has more than one purpose - it's a way to flag a language is in trouble and may get removed if people don't step up to maintain it, but it can also serve as a way to incubate new target languages.

If we have a new target language with a test suite in a good state and a maintainer willing to respond quickly to test suite failures, that's better than we have for at least half the target languages with "supported" status and it seems unhelpful to insist that test suite failures are marked to ignore as that makes them less likely to be promptly seen and attended to.  We already have to scan CI results and ignore failures due to CI system issues for a significant proportion of CI runs.

So I'm with @vadz on this one, with the proviso that it's a privilege to get this special treatment and that privilege is likely to be quickly revoked if test suite failures stop getting promptly addressed or occur with an annoying frequency.

wsfulton commented:
More than 10% of the test-suite is not running for C, so I see a fairly high probability of new testcases causing problems. It would indeed be a special privilege to get this treatment and would need to be earned. We don't really know much about this module during the normal development cycle, so let's revisit 3 months after the 3 merge criteria are fulfilled to see if it is worthy of special privilege given track record over the 3 months. I'd be supportive if there are just a handful of C only breakages that were promptly addressed (after being notified of a failure). Don't worry about them not getting noticed, I'll see them.

PS, I've got a list of improvements for this module that I made during the review that I'll be committing once the merge has been done. Might help a bit make it more mature and stable.

vadz commented:
Sorry, I don't understand: do you want me to allow failures for C backend builds and prefer to manually monitor them rather than rely on me actually being notified about the failures and fixing them myself? If so, why not, I guess, but I just can't fathom why would you want to do it, so maybe I'm misunderstanding you.

Also, concerning the "3 merge criteria", I still don't know what exactly must be changed for the docs, could you please clarify this?

Finally, concerning your improvements, why not push them to this branch right now rather than waiting for the merge?

wsfulton commented:
As the maintainer of the module, please improve the docs in areas that you feel are necessary, for example, you've identified areas to fix: 

> _I'm not too happy with its existing organization_.

>_Concerning the documentation, I was thinking of giving more examples of SWIG input and output for this backend

My comments so far:
> Your suggestion of giving more examples of SWIG input and output for this backend sounds great. Python and Java chapter are our best role models for converting to C, but consider that a longer term goal. I think the best thing would be to read through the current docs and fix what you know is incorrect or poor in the first instance.

That was a pretty open-ended, but as requested here are a more specific set of improvements to make:

1. First I suggest reading the entire chapter and correct the things which are plain wrong and turn it into something that you as the module maintainer are proud of.
2. Explain the thinking, motivation and differences wrt C output and C++ output in an introducton.
3. Perhaps restructure into two chapters or sub-sections, one for each of C and C++. The examples should consider both outputs. This is my main area of confusion and the lack of docs is one of the reasons for this.
4. Define an alternative name for 'proxy' and use it throughout the docs.
5. Copy Java's "A tour of basic C/C++ wrapping" section and modify. The sub-sections there cover the core of what needs to be documented.
6. Add a section on exceptions and how to handle them as this is an important area to cover.
7. Typemaps section - provide practical examples of how to use all the C specific typemaps.

> Finally, concerning your improvements, why not push them to this branch right now rather than waiting for the merge?

I'm now trying to focus on getting swig-4.1.0 out the door before getting too involved in this module. Also, I like to make changes and document those changes as I go along, but I can't because I still don't understand a lot of this module due to the poor state and structure of the docs. I really need your help to get a decent starting point.

> Sorry, I don't understand: do you want me to allow failures for C backend builds and prefer to manually monitor them rather than rely on me actually being notified about the failures and fixing them myself? If so, why not, I guess, but I just can't fathom why would you want to do it, so maybe I'm misunderstanding you.

The 'continue-as-error' is there so that it does not impact development of the supported language modules by breaking CI for contributor's pull requests nor commits made to master. It is a requirement for the experimental languages on master so that it can be easily improved towards supported status without impacting the normal development of the supported modules.

From file:///home/william/swig/github/swig/Doc/Manual/Extending.html#Extending_experimental_status
> If a language module has an official maintainer, then the maintainer will be requested to focus on fixing test-suite regressions and commit to migrating the module to become a 'Supported' module.

So, experimental module maintainers are expected to monitor the builds while the module is worked on towards completion to full status. So in answer to your question "do you want me to allow failures for C backend builds and prefer to manually monitor them", the answer is absolutely yes for the privilege of developing in master, that's wat continue-as-error requires. As I'm always looking at the builds, I am offering to give some assistance here by letting you know if there the C module goes red. Not sure if that answers, hopefully helps.

vadz commented:
Sorry but let me just confirm that I understand you correctly:

1. You don't plan to merge this in 4.1.0, so it's not going to be merged for at least a couple more years.
2. You require completely rewriting the documentation and this has to be done right now, before this is merged.
3. You require me to manually monitor all the CI failures in all the PR builds for the observable future and getting notified about the failures automatically is not an option.

Is this right? If it is, I'm afraid I'll have to decline the privilege of developing in master and will just keep this in my own branch, thank you. Please feel free to close this PR.

wsfulton commented:
> You don't plan to merge this in 4.1.0, so it's not going to be merged for at least a couple more years.

No. If it wasn't clear, I won't be adding changes I had in mind (these are not required for merging to master) as I'd like to focus on 4.1.0. I assure you that I will make a special effort to merge this pull request into master as soon as it is ready and without delay, independent of the 4.1.0 release and independent of additional changes I might make.

> You require completely rewriting the documentation and this has to be done right now, before this is merged.

I didn't request a rewrite and I don't really see that one is needed. I've provided open ended suggestions as to how to improve the documentation and when asked provided more detailed suggestions. The documentation does need to be improved though, absolutely. Hopefully I'm wrong, but there seems to be some reticence in even starting this. Let it be crystal clear though, I believe the documentation is as important as the code and until there is some improvement in the docs, I think this module is not acceptable.

> You require me to manually monitor all the CI failures in all the PR builds for the observable future and getting notified about the failures automatically is not an option.

Sorry, it seems the more I try and explain the more I might be confusing you, so I suggest you familiarise yourself with how `continue-as-error` works in Github and particular our usage of it with the other SWIG experimental languages. Also Github has a wide variety of automatic notification options, but sorry, I'm no expert on them.

vadz commented:
I am afraid I have no illusions that this is ever going to get merged any longer but, just in case I'm wrong, I've updated the PR after merging it with the latest master and fixing a few simple problems that were uncovered by the tests added to master.

BTW, if anybody else wants to try this branch out, please feel welcome to do so. There are definitely still many features that are not supported yet, but we've been using it for a year for a rather big C++ library without any problems.

jyapayne commented:
@vadz are you able to share the project you've been using it with? I would love to read the source code

gogo2464 commented:
@vadz should it be merged this year?

vadz commented:
> @vadz should it be merged this year?

It doesn't depend on me but I wouldn't hold your breath.

ojwb commented:
I've merged master to the branch, resolved the conflicts, and fixed a few things.

I updated the list of "known to fail" testcases so that `make check-c-test-suite` passes.  Many of the new failures were new or changed test cases, but a few were unchanged and pass on the unmerged branch which seems a bit suspicious (before the merge all test cases pass for me except `default_arg_expressions` - this one passes after the merge).

vadz commented:
> I've merged master to the branch, resolved the conflicts, and fixed a few things.

Thanks for doing this!
 
> I updated the list of "known to fail" testcases so that `make check-c-test-suite` passes. Many of the new failures were new or changed test cases, but a few were unchanged and pass on the unmerged branch which seems a bit suspicious (before the merge all test cases pass for me except `default_arg_expressions` - this one passes after the merge).

I could look into this, but I admit that motivation is difficult to find as the last time this was discussed I got a clear impression that there was no interest at all in merging this in observable future from William and, as long as this is not going to happen, why bother...

ojwb commented:
It does appear to be essentially impossible to get a new target language merged now (at least Fortran seems to be in a similar limbo despite a significant amount of work going into it).  The last genuinely new target language added seems to be D, and that was in 2010 (03aefbc6e95d094a6de231e1f5264c0946e209a3).  Before that Go earlier in 2010, Scilab in 2009, Octave in 2008, and CFFI, Lua, UFFI & Clisp (all in 2005).  So we were averaging about one new target language a year.  It's plausible we might expect that to tail off as the more that get added the more we already supports most of the languages people might want to add, but there have been none for over 13 years.

(I'm ignoring PHP7 which was technically a new target language backend, but that was essentially me forking the existing PHP5 target language because the C extension API had changed so much that trying to use conditional code would have been a nightmare; once PHP5 was out of support I nuked the entire PHP5 target language backend so the duplication was fairly short lived.)

I don't really know how to fix this blockage.  I don't feel it's useful for me to start merging new target languages when William's asked for further changes (and he's shown he's quite willing to revert my commits without any prior discussion).

We now have the 3 tiers of target languages so in theory it ought to be achievable to get a new target language in at the lowest tier, and then work on improving it without the on-going pain of merge conflicts from changes on git master.  Maybe this means the current requirements for the lowest tier are too strict and we need more of an incubator tier as well or instead.

ojwb commented:
I was just checking the current requirements, and actually the lowest tier is "Disabled" and is for existing target languages which will be deleted if someone doesn't fix them, so really there are currently only two tiers that are relevant for a new target language.

This branch seems like it satisfies the criteria for "Experimental" to me though.

gogo2464 commented:
I have an app that MUST be used in C as well as c++. Should I continue to write it in C with %extends swig keyword at this moment or should I do it in native C++ with C wrapper please?

ojwb commented:
@gogo2464 A C API + `%extend` is the only one of your two approaches that's currently supported by released versions of SWIG, so that seems the better option until this branch gets merged, which sadly may never happen.

gogo2464 commented:
@ojwb thanks! do you have a link on how to create c++ wrapper on C code? the -c++ arg is used to enable c++ input.

ojwb commented:
That's not really on topic for this ticket, but see https://www.swig.org/Doc4.2/SWIG.html#SWIG_adding_member_functions (and probably you want to use `-c++` even though the code being fed in is C).

gogo2464 commented:
the cicd seems to be fixed. Can I already use it right now for my project or should I wait please?

vadz commented:
While working on other SWIG-related issues, I've merged the latest master into this branch and updated it for cdata and isvoid typemap changes (neither of which is really useful for C, but at least the tests pass).

This being said, I don't know how much sense does it make to keep this branch up to date here, to be honest, if there is no chance of it ever being merged, perhaps it would be better to just create a swig-c fork instead?

ojwb commented:
> perhaps it would be better to just create a swig-c fork instead?

Switching to a model where any new target language has to be a fork of SWIG seems a terrible outcome to me.

I think we need to get back to the situation where it's actually possible to get new a target language in.

A viable new candidate target language needs to be at a point where we can see it basically works, so reasonable test coverage which passes with any failing testcases listed, and tests run in CI.  Once it is at that point it's simpler to have it merged both those working on the target language (as they don't need to keep merging from master and resolving conflicts) and those reviewing changes (because they only need to review the incremental changes).  The "Experimental" status in theory allows for this, but in practice (as I noted above) we've not merged a new target language since 2010, whereas we were averaging about one a year up to then.

Looking at the current state here, it really seems to match or surpass all the criteria for "Experimental" status as documented in the manual, and I think it's ready for the next step so I'm putting this on the 4.3 milestone.

If it's falling short in some way(s) I've missed, I think we need a clearly articulated list of what still needs resolving before it can be merged.

vadz commented:
> > perhaps it would be better to just create a swig-c fork instead?
> 
> Switching to a model where any new target language has to be a fork of SWIG seems a terrible outcome to me.

I obviously completely agree with this, but I don't know what else can be done.

ojwb commented on configure.ac:
```diff
@@ -2611,6 +2640,13 @@ AC_SUBST(TCLLINK)
 #----------------------------------------------------------------
 
+SKIP_C=
+if test -x "$CC" || test -z "$CXX" ; then
```
Should this `test -x` be `test -z`?

`-x` seems wrong here as this currently seems like it will skip C testing if `$CC` is executable, which I'd expect it always will be (since if we don't have a working C compiler, configure should bail out with an error).

I checked the last CI run and we actually get away with this in CI because `CC` doesn't have a full path there, and `test -x gcc` is not true because `test -x` doesn't try to resolve via `PATH`.  I do think it's wrong though.

vadz replied:
This definitely seems wrong, thanks for noticing. Apparently it always was like this, ever since a13c5976b (Fixed naming convention issues (internal Swig variables, shadow -> proxy, test.c -> runme.c, etc.), modified Makefiles, bugfixes ("name" -> "sym:name"). Started C test-suite., 2008-07-09), but this doesn't mean anything, of course.

AFAICS this entire fragment could be just removed completely because we require both C and C++ compiler.

ojwb replied:
> AFAICS this entire fragment could be just removed completely because we require both C and C++ compiler.

Indeed - I've merged master to the branch (no manual intervention needed) and have removed the check - `SKIP_C` is now always empty (but still present in case we later grow dependencies here, e.g. we might conceivably start using some C test harness in the future).

ojwb commented:
I'm unclear why it has stalled at this point.  I just looked over the current diff and only noticed a minor issue with the configure test (see comment on the diff - happy to push a fix, but I thought I shouldn't do that unilaterally).

If it's just needing someone to do the merge, I can try to find time.  If there are still remaining blockers before it can be merged, they really need to be articulated.  It's on the milestone for the next release now at least.

vadz commented:
> I'm unclear why it has stalled at this point.

In short, I don't really know. I'd very much like to, though.

gogo2464 commented:
if you need testing,I can test your code. If you want.

ojwb commented on Source/Modules/c.cxx:
```diff
@@ -0,0 +1,3149 @@
+    String *name = Getattr(n, "sym:name");
+    // If it's a #define or a %constant, use raw value and hope that it will work in C as well as in C++. This is not ideal, but using "value" is even worse, as
+    // it doesn't even work for simple char constants such as "#define MY_X 'x'", that would end up unquoted in the generated code.
+    String *value = Getattr(n, "rawval");
```
Once #2990 is merged, this will need updating.

The `rawval` attribute is gone, and `value` will be a valid C/C++ expression (so for a string or char literal it's suitably escaped and includes quotes).  For cases where we want to emit a string or character literal in the target language there's now `stringval` with the byte sequence for the string's value, but I think for C just using `value` should work in all cases we can actually reasonably handle.

I think consulting `staticmembervariableHandler:value` as below shouldn't then be necessary as there shouldn't be a situation where the `value` attribute isn't set but `staticmembervariableHandler:value` is.

So probably the line above (and the comment above it) and both the `if`s below can be replaced with just:

```
    String *value = Getattr(n, "value");
```

ojwb replied:
Done, along with another fix to stop escaping `T_CHAR` values.

vadz replied:
Thanks Olly! I did plan to do it when merging the latest master into our branch but hadn't managed to do it before you did, thanks again!

ojwb commented:
I've done another pass over the changes here and fixed a few minor niggles I noticed, and merged to master.  The changes here seem to unarguably fulfil all the documented requirements for an experimental target language (it actually seems fairly close to fulfilling the requirement for being a fully supported target language).

This won't give things much time to settle before the 4.3.0 release, but there are very few changes made to generic code.  If we don't include this in 4.3.0 it would probably be another year or more before this sees a release, whereas if we put it out in a release users can more easily try it and we can start getting wider testing.

gogo2464 commented:
Amazing ! I need to test!!!! Greatly thanks! @vadz @ojwb
